### PR TITLE
Refactor CLI flags, controller structure and controller configs

### PR DESCRIPTION
**Pull request by shreyas-s-rao**
**How to categorize this PR?**

/area quality
/area dev-productivity
/area testing
/area usability
/kind cleanup

**What this PR does / why we need it**:
This PR brings the following refactoring changes to etcd-druid:
-  refactors the CLI flags for better readability and testability, since the flags will now printed upon code execution. This makes it easier to figure out what configuration druid is running with.
- refactors the controller structure, by adding each controller as an individual package under `controllers`. The PR also adds ControllerConfigs per controller for better passing of controller configurations to each controller, and splits each controller code into 3 files:
  - `add.go` - adds the controller to the manager
  - `config.go` - provides controller configuration and initialisation via CLI flags
  - `reconciler.go` - provides reconciler and associated functionality
- `main.go` becomes much leaner, by not worrying about setting up the manager or adding each controller to it (this is moved into the `controllers` package now, which creates the manager and adds all underlying controllers to the manager
- unit tests for controllers are now focused only on testing functional units within controller code, while entire controllers are tested individually as part of the new `test/integration/*` test suites. 
  > NOTE: Currently the scope of unit tests is to only test helper functions/methods. Integration/unit tests are not comprehensive today and we need to address this via a separate issue/PR. We do not wish to increase the scope of this PR any further. We will create a series of separate dedicated issues which will be taken up to address all shortcomings that we have found while refactoring the code.

Tasks:
- [x] move reconciler logic from `controllers` into individual packages:
  - [x]  `controllers/etcd`
  - [x] `controllers/custodian` - merge with `controllers/controller_ref_manager.go` and remove unused methods
  - [x] `controllers/compaction`
  - [x] `controllers/etcdcopybackupstask`
  - [x] `controllers/secret`
- [x] write unit tests for each controller
  - [x]  `controllers/etcd` ❌ no unit tests identified to be added
  - [x] `controllers/custodian` ❌ no unit tests identified to be added
  - [x] `controllers/compaction` ❌ no unit tests identified to be added
  - [x] `controllers/etcdcopybackupstask`
  - [x] `controllers/secret`
- [x] Write unit tests for functions in `pkg/utils` package 
- [x] move controller tests into individual packages under `test/integration/controllers`:
  - [x]  `controllers/etcd`
  - [x] `controllers/custodian`
  - [x] `controllers/compaction`
  - [x] `controllers/etcdcopybackupstask`
  - [x] `controllers/secret`
- [x] Add make target for integration tests (make test-integration) and add to CI pipeline via new step `.ci/test_integration`
- [x] Remove `pkg/config` and `pkg/flags`
- [x]  Remove usage of deprecated pointer methods
- [x] Re-organize files under `api/v1alpha1`
- [x] Convert helper functions from `pkg/utils/names.go` into methods on `Etcd` struct
- [x] Run `make add-license-headers` for all newly added files
- [x] Ensure all tests are passing (unit, integration, e2e)
- [x] Run manual tests to confirm correct behaviour of all controllers

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/etcd-druid/issues/512

**Special notes for your reviewer**:

**Release note**:

```other developer
Refactor etcd-druid code: refactor CLI flags, restructure controllers, add controller configs, add controller-specific unit tests, restructure existing controller tests into controller-specific integration tests. This makes etcd-druid easier to maintain and contribute to.
```
```breaking user
The default value for druid CLI flag `ignore-operation-annotation` is now set to `false`. This means druid will now respect the `gardener.cloud/operation` annotation for reconciling the `Etcd` resource by default. This change allows for developers to follow best practices during local development and testing.
```
```feature developer
Introduce integration tests for testing the functionalities of each controller, which can be run using `make test-integration`.
```
```bugfix operator
Squash multiple bugs in controllers as part of the refactoring process.
```

### CODE REVIEW

File: main.go
Comment (unmarshall) [inline main.go:37]: Minor comment. Can you create 2 constant blocks - one for defaults and another for flag names?
Another minor comment (this is obviously a matter of personal choice) - I would prefer `flag` as a suffix rather than as a prefix. But this you can accept or reject. Whatever convention we follow then lets stick to that convention throughout.

Comment (shreyas-s-rao) [reply main.go:37]: Sure Madhav, will have two separate constant blocks. But naming-wise, I would prefer to keep it the same since the prefix `flag` or `default` denotes the purpose of the constant. So all `flag`-prefixed constants denote flag names, while all `default`-prefixed constants denote default values for the flags. Hope that's fine.

**Comment by shreyas-s-rao** [comment]
@unmarshall thanks for your review. I've address your comment. PTAL.
/merge squash

**Comment by unmarshall** [comment]
Another way to specify defaults is
https://github.com/kubernetes/kubernetes/blob/45df8f0bb3c0165c5a3e716e803402fa0bd7087d/cmd/kubeadm/app/apis/kubeadm/v1beta3/defaults.go
This way you could define functions that will give a default Leader election config or default Storage config etc.

**Comment by shreyas-s-rao** [comment]
@unmarshall thanks for your suggestion. I've now created two new packages:
- `pkg/config` which holds configuration defaults and provides config structs for each of the controllers and the controllerManager, as well as potentially webhooks and other config in the future. Multiple files make it easy to differentiate between configs of different components of druid.
- `pkg/flags` which holds CLI flag names and provides flag-related functionality like adding, parsing and printing flags.

This makes `main.go` quite lean (although we have potential to make it leaner, which we can do in a separate refactoring story), and also makes it easy to add new components to druid, such as various webhooks (which will be required in the coming days).

PTAL

File: pkg/config/controller_manager.go
Comment (unmarshall) [inline pkg/config/controller_manager.go:51]:
```suggestion
func DefaultControllerManagerConfig() ControllerManagerConfig {
```

File: pkg/config/custodian_controller.go
Comment (unmarshall) [inline pkg/config/custodian_controller.go:45]:
```suggestion
func DefaultCustodianControllerConfig() CustodianControllerConfig {
```

File: pkg/config/etcd_controller.go
Comment (unmarshall) [inline pkg/config/etcd_controller.go:28]:
```suggestion
func DefaultEtcdControllerConfig() EtcdControllerConfig {
```

File: pkg/config/etcdcopybackupstask_controller.go.go
Comment (unmarshall) [inline pkg/config/etcdcopybackupstask_controller.go.go:28]:
```suggestion
func DefaultEtcdCopyBackupsTaskControllerConfig() EtcdCopyBackupsTaskControllerConfig {
```

File: pkg/config/secret_controller.go
Comment (unmarshall) [inline pkg/config/secret_controller.go:28]:
```suggestion
func DefaultSecretControllerConfig() SecretControllerConfig {
```

File: pkg/config/controller_manager.go
Comment (unmarshall) [inline pkg/config/controller_manager.go:33]: These are command line flags and I would personally prefer to follow well accepted structure where we have a top level `cmd` directory instead where we have `LeaderElectionConfig` and likewise defined. `main.go` will anyways have a dependency to commands.

File: pkg/config/compaction_lease_controller.go
Comment (unmarshall) [inline pkg/config/compaction_lease_controller.go:27]: Instead of creating a config package. I would recommend that we create one per functionality. This will later help in moving parts of the controller into this folder and making the controllers more leaner.

File: pkg/flags/flags.go
Comment (unmarshall) [inline pkg/flags/flags.go:91]: There is no need to expose these methods if they are only going to be used from within `AddAllFlags`
Same comment applies to all other similar methods.

File: pkg/flags/flags.go
Comment (unmarshall) [inline pkg/flags/flags.go:112]: This entire thing should belong in `cmd` package and not here.

**Comment by unmarshall** [comment]
In general i would say keep this PR limited to only changing what you need. If you wish to the do a larger refactoring then lets have a pair programming session.

**Comment by shreyas-s-rao** [comment]
Rebased on latest master branch: [diff](https://github.com/gardener/etcd-druid/compare/32c2377889b50b5456d0d20a875d024356158e60..6a93d1cfcf4776756af4b115f38d6971e16f6182)

**Comment by shreyas-s-rao** [comment]
/invite @ishan16696 @abdasgupta @aaronfern @timuthy 
PTAL

**Comment by shreyas-s-rao** [comment]
[make-test.log](https://github.com/gardener/etcd-druid/files/10716934/make-test.log)
[make-test-integration.log](https://github.com/gardener/etcd-druid/files/10716939/make-test-integration.log)

As you can see, previous test suite needed 20-25 minutes to run. The refactored unit test suite and integration test suite need only ~25 seconds and ~7.5 minutes respectively to finish execution. 

(Please check results of `make test-integration` from the attached logs, since it will only be run in the CICD pipeline *after* this PR is merged)

**Review by aaronfern — CHANGES_REQUESTED** [review]
Thanks @shreyas-s-rao and @unmarshall for this well structured PR!
A few comments from me

File: controllers/config.go
Comment (unmarshall) [reply controllers/config.go:56]: This is struct embedding. If you see `InitFromFlags` and see how the fields within `LeaderElectionConfig` are used you will realize.

File: api/v1alpha1/types_common.go
Comment (unmarshall) [reply api/v1alpha1/types_common.go:58]: Yes it should be ConditionStatus. However both `ConditionProgressing` and `ConditionCheckError` are never used. For now we have not changed what was there earlier and only re-arrange the code. But to be consistent we can add the proper type. Thanks for pointing that out.

File: controllers/compaction/config.go
Comment (unmarshall) [reply controllers/compaction/config.go:54]: In general you are right. In the earlier code (current master version) we do not have any validations. So we can add validations where required.

File: api/v1alpha1/types_common.go
Comment (shreyas-s-rao) [reply api/v1alpha1/types_common.go]: Agreed. Anyway this constant isn't used anywhere and will be removed in a future API change

File: controllers/compaction/config.go
Comment (shreyas-s-rao) [reply controllers/compaction/config.go]: I have added validations for all controller configs and controller manager config too. Thanks for bringing this up @aaronfern !

File: controllers/compaction/reconciler.go
Comment (shreyas-s-rao) [reply controllers/compaction/reconciler.go:197]: From the [existing code](https://github.com/gardener/etcd-druid/blame/d29781f8270e020d244d94df75080163d85cea6e/controllers/compaction_lease_controller.go), you can see that there have been some inconsistencies in the delete method and the way it's called. This PR fixes those inconsistencies.

**Comment by shreyas-s-rao** [comment]
@aaronfern thanks for your review. @unmarshall and I have addressed your comments. PTAL

File: api/v1alpha1/types_etcd_test.go
Comment (abdasgupta) [inline api/v1alpha1/types_etcd_test.go:67]: As we are refactoring, I think it's a good idea to use `context.WithTimeout()` and use a constant timeout value all through the code

File: controllers/create.go
Comment (abdasgupta) [inline controllers/create.go:1]: Can we rename the file to manager.go ? It's mostly related to handling manager, ryt?

File: controllers/custodian/add.go
Comment (abdasgupta) [inline controllers/custodian/add.go:1]: We can again rename the file to manager.go

File: controllers/secret/reconciler_test.go
Comment (abdasgupta) [inline controllers/secret/reconciler_test.go:1]: Why is this test here? Why is it not placed with the tests for other controller?

**Review by abdasgupta — CHANGES_REQUESTED** [review]
Thanks for this PR. I have some minor comments

File: controllers/secret/reconciler_test.go
Comment (unmarshall) [reply controllers/secret/reconciler_test.go:1]: Tests that are placed within `controller/<specific-controller>` are only unit tests. As and when more unit tests are identified for every controller it should be placed within the same package. Only integration tests will go to `test/integration`

File: controllers/custodian/add.go
Comment (unmarshall) [reply controllers/custodian/add.go:1]: ok, sure

File: controllers/create.go
Comment (unmarshall) [reply controllers/create.go:1]: yes that would be fine. This is where the manager is created and all reconcilers are also added.

File: controllers/custodian/add.go
Comment (unmarshall) [reply controllers/custodian/add.go:1]: In this file we will only have functions which will add a reconciler to the manager. We have borrowed this convention from g/g. Out of several examples of such usage there, you can see one usage at https://github.com/gardener/gardener/tree/master/pkg/controller/service

Comment (abdasgupta) [reply controllers/custodian/add.go:1]: register.go seems fine

File: controllers/create.go
Comment (unmarshall) [reply controllers/create.go]: Done

File: controllers/compaction/config.go
Comment (shreyas-s-rao) [inline controllers/compaction/config.go:62]: Workers can be zero too, which means the controller will be disabled. Since compaction controller is an "extension" or non-essential controller for the functioning of etcd-druid, it can be enabled/disabled as necessary. That's the reason I had previously allowed `cfg.Workers` to take a zero-value as well.

```suggestion
	if err := utils.MustBeGreaterThan(workersFlagName, -1, cfg.Workers); err != nil {
```

File: controllers/etcdcopybackupstask/config.go
Comment (shreyas-s-rao) [inline controllers/etcdcopybackupstask/config.go:43]: Same as above.

File: controllers/config.go
Comment (shreyas-s-rao) [inline controllers/config.go:136]: `TODO`s make it easier to follow-up on technical debts.

```suggestion
// TODO: This should be changed as lease is the default choice now for leader election. There is no need
```

**Review by shreyas-s-rao — COMMENTED** [review]
@unmarshall thanks a lot for adding support for generics in controller flag validation. I wanted to point out couple of things in the changed code that deviates fro the original changes. PTAL

File: controllers/custodian/add.go
Comment (shreyas-s-rao) [reply controllers/custodian/add.go:1]: "add" and "register" are synonymously used terms for the same operation - adding a controller to a controller manager. We borrowed the term "add" from g/g (ex: https://github.com/gardener/gardener/tree/master/pkg/controllermanager/controller/bastion) which does the same thing as we're doing here.

If we decide to rename to "register.go", it would then be required to rename the method `Reconciler.AddToManager()` to `Reconciler.RegisterWithManager()` for all the controllers as well, for semantic correctness. @unmarshall WDYT?

File: api/v1alpha1/types_etcd_test.go
Comment (shreyas-s-rao) [reply api/v1alpha1/types_etcd_test.go:67]: This is possible, although we will need to use different timeout values for each kind of request and resource. For instance, the time required to get a secret may be significantly lesser than the time required to create an `Etcd` resource.

File: controllers/custodian/add.go
Comment (unmarshall) [reply controllers/custodian/add.go:1]: That is the reason i have not yet changed it. I have personally no problem with the name `add.go` as it goes well with `AddToManager`

File: api/v1alpha1/types_etcd_test.go
Comment (shreyas-s-rao) [reply api/v1alpha1/types_etcd_test.go:67]: @abdasgupta we plan to do all test-specific refactoring as part of a separate PR. See https://github.com/gardener/etcd-druid/issues/527 for follow-up. So making further changes to the tests is out of scope for this PR. You may also check the `NOTE` from @unmarshall in this PR's description.

Comment (unmarshall) [reply api/v1alpha1/types_etcd_test.go:67]: If you do not mind, can we do this as part of another PR? We have already increased the scope of this PR by a lot (from where we started).

**Comment by ishan16696** [comment]
Hi @shreyas-s-rao @unmarshall ,
thanks for all refactoring.
I observed that you have change the `etcdCopybackupController`, Have you also tested the copy backup task with `etcdCopybackupController` ? If not then I will do it. Please let me know.

File: controllers/compaction/config.go
Comment (unmarshall) [reply controllers/compaction/config.go:62]: I wish we use flags to enable/disable controllers and not workers. For now i will provide another method `MustBeGreaterThanOrEqualTo`

File: controllers/config.go
Comment (unmarshall) [reply controllers/config.go]: done

File: controllers/etcdcopybackupstask/config.go
Comment (unmarshall) [reply controllers/etcdcopybackupstask/config.go]: check the changes.

File: controllers/compaction/config.go
Comment (unmarshall) [reply controllers/compaction/config.go]: check the changes in the latest commit

File: api/v1alpha1/types_common.go
Comment (unmarshall) [reply api/v1alpha1/types_common.go:32]: In this PR we have not looked at changing the API. This remains the same as it was before. There will be another PR where we will make changes to the API. Perhaps we can revisit your comment then. Ok with you?

File: config/samples/druid_v1alpha1_etcd.yaml
Comment (unmarshall) [reply config/samples/druid_v1alpha1_etcd.yaml:76]: In this PR we have not touched this part at all and remains out of scope of this PR. However, just an observation - we support both a single node etcd cluster and a multi-node etcd cluster. Since this is a sample, we have to choose one value. The consumer can easily change the replicas to be equal to an acceptable odd number.

File: controllers/custodian/add.go
Comment (unmarshall) [reply controllers/custodian/add.go:1]: @shreyas-s-rao - 2 reviewers prefer `RegisterWithManager` and `register.go`. I do not have any strong opinion on this and i am ok with either of the names. Should we then make this change?

File: config/samples/druid_v1alpha1_etcd.yaml
Comment (ishan16696) [reply config/samples/druid_v1alpha1_etcd.yaml:76]: I would suggest as this is a very small change we can do this in this PR.
> Since this is a sample, we have to choose one value. The consumer can easily change the replicas to be equal to an acceptable odd number.

if `spec.replicas` is `3` then it seems obvious that `1` will also be supported but if `spec.replicas` is `1` in sample then it may not be implicit that it supports `3`(>1) . Atleast this is how I see it but I leave it to you to decide.

File: controllers/custodian/add.go
Comment (shreyas-s-rao) [reply controllers/custodian/add.go:1]: Ok, in that case let's make the changes.

File: config/samples/druid_v1alpha1_etcd.yaml
Comment (shreyas-s-rao) [reply config/samples/druid_v1alpha1_etcd.yaml:76]: Sure, we can progress the sample etcd to 3 replicas in this PR.

**Comment by shreyas-s-rao** [comment]
@abdasgupta @ishan16696 I have addressed your review comments. PTAL
cc @unmarshall

**Review by abdasgupta — APPROVED** [review]

**Review by ishan16696 — APPROVED** [review]

**Review by unmarshall — APPROVED** [review]
/lgtm

**Review by aaronfern — APPROVED** [review]

**Pull request merged**
